core/src/main/java/io/neow3j/transaction/Transaction.java

```diff
@@ -0,0 +1,463 @@
+    }
+
+    public void addWitness(Witness witness) {
+        if (witness.getScriptHash() == null || witness.getScriptHash().length() == 0) {
+            throw new IllegalArgumentException("The script hash of the given script is " +
+                    "empty. Please set the script hash.");
+        }
+        this.witnesses.add(witness);
+    }
+
+    public String getTxId() {
+        byte[] hash = Hash.sha256(Hash.sha256(toArrayWithoutWitnesses()));
+        return Numeric.toHexStringNoPrefix(ArrayUtils.reverseArray(hash));
+    }
+
+    public int getSize() {
+        return toArray().length;
+    }
+
+    @Override
+    public void deserialize(BinaryReader reader) throws DeserializationException {
+        try {
+            this.version = reader.readByte();
+            this.nonce = reader.readUInt32();
+            this.sender = reader.readSerializable(ScriptHash.class);
+            this.systemFee = reader.readInt64();
+            this.networkFee = reader.readInt64();
+            this.validUntilBlock = reader.readUInt32();
+            this.attributes = reader.readSerializableList(TransactionAttribute.class);
+            this.cosigners = reader.readSerializableList(Cosigner.class);
+            this.script = reader.readVarBytes();
+            this.witnesses = reader.readSerializableList(Witness.class);
+        } catch (IOException | InstantiationException | IllegalAccessException e) {
+            throw new DeserializationException(e);
+        }
+    }
+
+    private void serializeWithoutWitnesses(BinaryWriter writer) throws IOException {
+        writer.writeByte(this.version);
+        writer.writeUInt32(this.nonce);
+        writer.writeSerializableFixed(this.sender);
+        writer.writeInt64(this.systemFee);
+        writer.writeInt64(this.networkFee);
+        writer.writeUInt32(this.validUntilBlock);
+        writer.writeSerializableVariable(this.attributes);
+        writer.writeSerializableVariable(this.cosigners);
+        writer.writeVarBytes(this.script);
+    }
+
+    @Override
+    public void serialize(BinaryWriter writer) throws IOException {
+        serializeWithoutWitnesses(writer);
+        this.witnesses.sort(Comparator.comparing(Witness::getScriptHash));
+        writer.writeSerializableVariable(this.witnesses);
+    }
+
+    /**
+     * Serializes this transaction to a raw byte array without any witnesses.
+     * <p>
+     * In this form, the transaction byte array can be used for example to create a signature.
+     *
+     * @return the serialized transaction
+     */
+    public byte[] toArrayWithoutWitnesses() {
+        try (ByteArrayOutputStream ms = new ByteArrayOutputStream()) {
+            try (BinaryWriter writer = new BinaryWriter(ms)) {
+                serializeWithoutWitnesses(writer);
+                writer.flush();
+                return ms.toByteArray();
+            }
+        } catch (IOException ex) {
+            throw new UnsupportedOperationException(ex);
```

Comment:
It would be nicer if some sort of SerializationException is thrown but this is how the `NeoSerializableInterface.toArray()` method does it. Introducing a checked `SerializationExceptions` would require try-catch blocks in many places.